trainer/finetune.py

```diff
@@ -0,0 +1,240 @@
+    "train": os.path.join(DATA_DIR, "rd-train.tsv"),
+    "validation": os.path.join(DATA_DIR, "rd-validation.tsv")
+}
+
+def tf_verbosity_level(level):
+  """Changes verbosity level."""
+  og_level = tf.logging.get_verbosity()
+  tf.logging.set_verbosity(level)
+  yield
+  tf.logging.set_verbosity(og_level)
+
+def _prediction_file_to_ckpt(path):
+  """Extract the global step from a prediction filename."""
+  return int(path.split("_")[-2])
+
+def load_predictions(task_name):
+  """Loads the most recent predictions in as ([(input, target, pred)], step)."""
+  # Grab the dataset for this task.
+  ds = t5.data.TaskRegistry.get(task_name).get_dataset(
+      split="validation",
+      sequence_length={"inputs": INPUT_LENGTH, "targets": TARGET_LENGTH},
+      shuffle=False)
+
+  # Grab the paths of all logged predictions.
+  prediction_files = tf.io.gfile.glob(
+      os.path.join(
+          MODEL_DIR,
+          "validation_eval/%s_*_predictions" % task_name))
+
+  # Get most recent prediction file by sorting by their step.
+  latest_prediction_file = sorted(
+      prediction_files, key=_prediction_file_to_ckpt)[-1]
+
+  checkpoint_step =_prediction_file_to_ckpt(latest_prediction_file)
+
+  # Collect (inputs, targets, prediction) from the dataset and predictions file
+  results = []
+  with tf.io.gfile.GFile(latest_prediction_file) as preds:
+    for ex, pred in zip(tfds.as_numpy(ds), preds):
+      results.append((tf.compat.as_text(ex["inputs_plaintext"]),
+                      tf.compat.as_text(ex["targets_plaintext"]),
+                      pred.strip()))
+  return (results, checkpoint_step)
+
+def save_metrics(task_name):
+  """Prints and saves metrics for the most recent checkpoint."""
+  results, checkpoint_step = load_predictions(task_name)
+  predictions = [line[2] for line in results]
+  targets = [line[1] for line in results]
+
+  hyp = list(map(lambda x: x.split(), predictions))
+  ref = list(map(lambda x: [x.split()], targets))
+
+  nltk_bs = nltk.translate.bleu_score.corpus_bleu(list_of_references=ref, hypotheses=hyp)
+  sb_bs = str(sacrebleu.corpus_bleu(predictions, [targets]))
+
+  print("NLTK BLEU SCORE: {:f}, SACREBLEU BLEU SCORE: {:s}, CHECKPOINT: {:d}".format(nltk_bs, sb_bs, checkpoint_step))
+  # Writes to $MODEL_DIR$/validation_eval/metrics$CHECKPOINT_NUMBER.json
+  metrics_path = os.path.join(
+          MODEL_DIR,
+          "validation_eval/metrics" + str(checkpoint_step) + ".json")
+  json.dump({"nltk_bleu_score" : nltk_bs, "sacrebleu_blue_score" : sb_bs, "recall@1" : 0}, tf.io.gfile.GFile(metrics_path, "w"))
+
+
+def main(argv):
+  """Main method for fintuning: builds, trains, and evaluates t5."""
+  tf.disable_v2_behavior()
+  tf.compat.v1.enable_eager_execution()
+
+  global PRETRAINED_DIR
+  global MODEL_DIR
+  PRETRAINED_DIR = os.path.join(BASE_PRETRAINED_DIR, FLAGS.size)
+  MODEL_DIR = os.path.join(MODELS_DIR, FLAGS.size, FLAGS.name)
+
+  print("--------DETECTING TPUs----")
+  try:
+    TPU_TOPOLOGY = "2x2"
```

Comment:
Does this need to be a constant or a flag? Will we be changing this if we want to fine-tune on top of bigger t5 models?

Reply:
I changed it to a flag, I'm not sure what else is available on our current tier of GCP. I think this will be able to handle the larger models but we can look into it.